File: ch8/UVa11212.cpp
// Editing a Book

#include <bits/stdc++.h>
using namespace std;

#define REP(i, n) for (int i = 0, _n = (n); i < _n; i++)

typedef map<int, int> mii;

int encode(vector<int> &a) { // turn array into a number for efficiency
  int ret = 0;
  REP (i, a.size())
    ret = (ret * 10) + a[i];
  return ret;
}

// BFS from s until nL depth, store the state distance in "dis",
// "ref" is a state distance from another BFS source.
// If a node exists in "ref" then return the "meet in the middle distance"
//    from source s and from another source (ref's source)
int bfs(int s, int nL, mii &dis, mii &ref) {
  queue<int> q; q.push(s);
  dis.clear(); dis[s] = 0;
  for (int L = 0; L < nL && q.size(); L++) { // limit to depth nL only
    REP (qq, q.size()) { // we do not use while (!q.empty()) to ensure we only touch nL layers only
      int u = q.front(), t = u; q.pop();
      vector<int> a;
      REP (i, 9) a.push_back(t%10), t /= 10; // unpack the digits again
      reverse(a.begin(), a.end());
      for (int i = 0; i < 9; i++) { // try all possible substring
        for (int j = i; j < 9; j++) {
          vector<int> b = a;
          b.erase(b.begin()+i, b.begin()+j+1); // cut index [i..j]
          for (int k = 0; k < b.size(); k++) { // try all possible paste point
            vector<int> c = b;
            for (int l = j; l >= i; l--) // paste at location k (using array a)
              c.insert(c.begin()+k, a[l]);
            int v = encode(c);
            if (ref.count(v)) return L + 1 + ref[v]; // touching the other side, this is the answer: L + 1 (this move) + from v to sorted state
            if (dis.count(v)) continue; // if revisited, ignore
            // relax and push to queue again
            dis[v] = L+1;
            q.push(v);
          }
        }
      }
    }
  }
  return 5;	// maximum distance (upperbound)
}

int main() {
  mii d[10];
  int n, caseNo = 1;
  for (n = 1; n < 10; n++) { // precalculate bfs from 1, 12, 123, ..., 123456789
    int s = 0; REP(i, n) s = s * 10 + i + 1; // the source
    bfs(s, min(2, n), d[n], d[0]); // only need to go 2 level deep, d[0] is blank at this stage
  }
  while (scanf("%d", &n) != EOF && n) {
    vector<int> a(n);
    REP(i, n) scanf("%d", &a[i]); // up to n = 9, 1 digit, can be encoded as integer
    int v = encode(a), res;
    if (d[n].count(v)) res = d[n][v];	// search from s = 123456789, if found, we have an answer
    else               res = bfs(v, 2, d[0], d[n]);	// otherwise, search from target v (2 level deep), if still not found, answer = 5
    printf("Case %d: %d\n", caseNo++, res);
  }
  return 0;
}

File: ch8/UVa11212.ml
(* Editing a Book *)
open Scanf
open Printf

(* module IntMap = Map.Make(struct type t = int let compare = compare end) *)

let encode arr =
  let ret = ref 0 in
  let n = Array.length arr in
  for i = 0 to n-1 do 
    ret := (!ret * 10) + arr.(i)
  done;
  !ret

let decode x =
  let a = Array.make 9 0 and t = ref x in
  for i = 8 downto 0 do 
    a.(i) <- !t mod 10;
    t := !t / 10
  done;
  a

let bfs s nL dis refr =
  let q = Queue.create () in
  Queue.push s q;
  Hashtbl.reset dis;
  Hashtbl.add dis s 0;
  let l = ref 0 in
  let ret = ref 5 in
  while !l < nL && not (Queue.is_empty q) do 
    for qq = 1 to Queue.length q do 
      let u = Queue.take q in
      let a = decode u in
      for i = 0 to 8 do 
        for j = i to 8 do 
          let b = Array.(append (sub a 0 i) (sub a (j+1) (8-j))) in
          let b_len = Array.length b in
          let slice = Array.sub a i (j-i+1) in
          for k = 0 to b_len-1 do
            let c = Array.(concat [ sub b 0 k; slice; sub b k (b_len-k) ]) in
            let v = encode c in
            if !ret = 5 && Hashtbl.mem refr v
            then begin
              ret := !l + 1 + Hashtbl.find refr v;
            end;
            if not (Hashtbl.mem dis v) then begin
              Hashtbl.add dis v (!l+1);
              Queue.push v q
            end
          done
        done
      done
    done;
    incr l
  done;
  !ret

let () = 
  let d = Array.init 10 (fun _ -> Hashtbl.create 100) and case = ref 1 in
  for n = 1 to 9 do 
    let s = ref 0 in
    for i = 1 to n do 
      s := !s * 10 + i;
    done;
    ignore (bfs !s (min 2 n) d.(n) (Hashtbl.create 1));
  done;
  let rec solve () =
    scanf "%d" (fun n ->
      if n <> 0 then begin
        let a = Array.make n 0 in
        scanf "\n%d" (fun x -> a.(0) <- x);
        for i = 1 to n-1 do 
          scanf " %d" (fun x -> a.(i) <- x)
        done;
        scanf "\n" ();
        let v = encode a and res = ref 0 in
        if Hashtbl.mem d.(n) v
        then
          res := Hashtbl.find d.(n) v
        else begin
          res := bfs v 2 d.(0) d.(n)
        end;
        printf "Case %d: %d\n" !case !res;
        incr case;
        solve ()
      end
    )
  in
  solve ()
